ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,211 @@
+    // then the remote healthy ones, and finally the possibly down ones.
+    List<ReplicaId> replicas = partitionId.getReplicaIds();
+    LinkedList<ReplicaId> downReplicas = new LinkedList<>();
+    Collections.shuffle(replicas);
+
+    for (ReplicaId replicaId : replicas) {
+      String replicaDcName = replicaId.getDataNodeId().getDatacenterName();
+      if (!replicaId.isDown()) {
+        if (replicaDcName.equals(datacenterName)) {
+          replicaPool.addFirst(replicaId);
+        } else if (crossColoEnabled) {
+          replicaPool.addLast(replicaId);
+        }
+      } else {
+        if (replicaDcName.equals(datacenterName)) {
+          downReplicas.addFirst(replicaId);
+        } else if (crossColoEnabled) {
+          downReplicas.addLast(replicaId);
+        }
+      }
+    }
+    replicaPool.addAll(downReplicas);
+    totalReplicaCount = replicaPool.size();
+    if (totalReplicaCount < successTarget) {
+      throw new IllegalArgumentException(
+          "Total Replica count " + totalReplicaCount + " is less than success target " + successTarget);
+    }
+    this.otIterator = new OpTrackerIterator();
+  }
+
+  @Override
+  public boolean hasSucceeded() {
+    return succeededCount >= successTarget;
+  }
+
+  @Override
+  public boolean isDone() {
+    return hasSucceeded() || hasFailed();
+  }
+
+  @Override
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
+    inflightCount--;
+    if (isSuccessFul) {
+      succeededCount++;
+    } else {
+      failedCount++;
+    }
+    getTracker(replicaId).update(time.milliseconds() - replicaRequestSendTimes.remove(replicaId));
+  }
+
+  @Override
+  public Iterator<ReplicaId> getReplicaIterator() {
+    replicaIterator = replicaPool.iterator();
+    return otIterator;
+  }
+
+  /**
+   * @return {@code true} if the operation has failed.
+   */
+  private boolean hasFailed() {
+    return (totalReplicaCount - failedCount) < successTarget;
+  }
+
+  /**
+   * Gets the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that {@code replicaId}
+   * belongs to.
+   * @param replicaId the {@link ReplicaId} whose request latency is going to be tracked.
+   * @return the {@link Histogram} that tracks requests to the class of replicas (intra or inter DC) that
+   * {@code replicaId} belongs to.
+   */
+  private Histogram getTracker(ReplicaId replicaId) {
+    if (replicaId.getDataNodeId().getDatacenterName().equals(datacenterName)) {
+      return localColoTracker;
+    }
+    return crossColoTracker;
+  }
+
+  /**
+   * An iterator to fetch replicas to send requests to. Respects parallelism and discounts requests that have been
+   * outstanding for a certain amount of time from parallelism.
+   */
+  private class OpTrackerIterator implements Iterator<ReplicaId> {
+
+    @Override
+    public boolean hasNext() {
+      return replicaIterator.hasNext() && (inflightCount < parallelism || isMostRecentRequestPastDue());
```

Comment:
Ah ok - with the refactoring this is still necessary (and derived from the super class)